byzcoin/service.go

```diff
@@ -1001,106 +1001,115 @@ func (s *Service) downloadDB(sb *skipchain.SkipBlock) error {
 	log.Lvlf2("%s: downloading DB", s.ServerIdentity())
 	idStr := fmt.Sprintf("%x", sb.SkipChainID())
 
-	// Loop over all nodes that are not the leader and
-	// not subleaders, to avoid overloading those nodes.
-	nodes := len(sb.Roster.List)
-	subLeaders := int(math.Ceil(math.Pow(float64(nodes), 1./3.)))
-	for ri := 1 + subLeaders; ri < nodes; ri++ {
-		// Create a roster with just the node we want to
-		// download from.
-		roster := onet.NewRoster(sb.Roster.List[ri : ri+1])
-
-		err := func() error {
-			// First delete an existing stateTrie. There
-			// cannot be another write-access to the
-			// database because of catchingLock.
-			_, err := s.getStateTrie(sb.SkipChainID())
-			if err == nil {
-				// Suppose we _do_ have a statetrie
-				db, stBucket := s.GetAdditionalBucket(sb.SkipChainID())
-				err := db.Update(func(tx *bbolt.Tx) error {
-					return tx.DeleteBucket(stBucket)
-				})
-				if err != nil {
-					return fmt.Errorf("Cannot delete existing trie while trying to download: %s", err.Error())
-				}
-				s.stateTriesLock.Lock()
-				delete(s.stateTries, idStr)
-				s.stateTriesLock.Unlock()
+	err := func() error {
+		// First delete an existing stateTrie. There
+		// cannot be another write-access to the
+		// database because of catchingLock.
+		_, err := s.getStateTrie(sb.SkipChainID())
+		if err == nil {
+			// Suppose we _do_ have a statetrie
+			db, stBucket := s.GetAdditionalBucket(sb.SkipChainID())
+			err := db.Update(func(tx *bbolt.Tx) error {
+				return tx.DeleteBucket(stBucket)
+			})
+			if err != nil {
+				return fmt.Errorf("Cannot delete existing trie while trying to download: %s", err.Error())
 			}
+			s.stateTriesLock.Lock()
+			delete(s.stateTries, idStr)
+			s.stateTriesLock.Unlock()
+		}
 
-			// Then start downloading the stateTrie over the network.
-			cl := NewClient(sb.SkipChainID(), *roster)
-			var db *bbolt.DB
-			var bucketName []byte
-			var nonce uint64
-			for {
-				// Note: we trust the chain therefore even if the reply is corrupted,
-				// it will be detected by difference in the root hash
-				resp, err := cl.DownloadState(sb.SkipChainID(), nonce, catchupFetchDBEntries)
-				if err != nil {
-					return errors.New("cannot download trie: " + err.Error())
-				}
-				if db == nil {
-					db, bucketName = s.GetAdditionalBucket([]byte(idStr))
-					nonce = resp.Nonce
-				}
-				// And store all entries in our local database.
-				err = db.Update(func(tx *bbolt.Tx) error {
-					bucket := tx.Bucket(bucketName)
-					for _, kv := range resp.KeyValues {
-						err := bucket.Put(kv.Key, kv.Value)
-						if err != nil {
-							return err
-						}
-					}
-					return nil
-				})
-				if err != nil {
-					return fmt.Errorf("Couldn't store entries: %s", err.Error())
-				}
-				if len(resp.KeyValues) < catchupFetchDBEntries {
-					break
-				}
+		// Create a roster with all nodes other than us
+		// TODO: remove this and use the future parallel client
```

Comment:
are you using the parallel client? the PR says it depends on it but I don't see where the parallel client is used

Reply:
This is in #2001

Reply:
The adding of `SendProtobufParallel` was much bigger than thought, as I had to adjust a couple of tests to the new parallelized calls. Now I don't have any `SendProtobufParallel` in it anymore, so I removed it from the description...